### ISSUE TITLE
Handle master node pod-cidr better

### ISSUE THREAD

**Issue by bprashanth** [kubernetes/kubernetes#32900]
Figure out the requirements and make it more dynamic, see discussion on: https://github.com/kubernetes/kubernetes/pull/32886, it's also caused issues before (https://github.com/kubernetes/kubernetes/pull/27202, https://github.com/kubernetes/kubernetes/pull/27016, https://github.com/kubernetes/kubernetes/issues/29756). I actually think the nodecontroller allocates a cidr to the master, we can probably use that:

``` console
$ kubectl get no kubernetes-master --template={{.spec.podCIDR}}
10.244.0.0/24
```

**Comment by roberthbailey** [comment]
If the master node is registered to the cluster, it shouldn't be treated any different from a regular node (modulo taints / unschedulability).

**Comment by bprashanth** [comment]
@kubernetes/sig-node @kubernetes/sig-network  @freehan what's the expected behavior when there's a conflict between pod-cidr and node.podCIDR?     

1. Does Kubenet handle changing cidr (we decided against this for changes to node.podCIDR, most of those arguments apply here too https://github.com/kubernetes/kubernetes/issues/19854)
2. Do other CNI plugins handle changing CIDR (there's going to be delay in static route propogation etc if CIDR changes)
3. Does kubelet handle changing podCIDR (i.e does the changed cidr even make it into CNI.conf?)

**Comment by freehan** [comment]
> Does Kubenet handle changing cidr (we decided against this for changes to node.podCIDR, most of those arguments apply here too #19854)

Once pod cidr is set, kubenet will ignore subsequent changes.

>Does kubelet handle changing podCIDR (i.e does the changed cidr even make it into CNI.conf?)

if `reconcile-cidr` flag is set, kubelet will pass the change to network plugins. 

> Do other CNI plugins handle changing CIDR (there's going to be delay in static route propogation etc if CIDR changes)

For `cni` network plugin, it never takes node.podCIDR directly. It reads cni config files from directories.

**Comment by bprashanth** [comment]
That's what I thought. So basically to get the behavior described elsewhere (start containers with pod-cidr and then *restart* with node.podCIDR) we would need kubelet to recreate all containers on change after writing new CIDR to file, and kubenet/other CNI plugins to pick up the change through file.

I don't see the point of piping the new cidr down the the plugins and having the plugins ignore the change.
I argued along these lines in https://github.com/kubernetes/kubernetes/issues/19854, where we basially decided to ignore the changing CIDR case, but maybe there's an exception if the change goes from --pod-cidr to node.podCIDR.

**Comment by gmarek** [comment]
NodeController respects CIDRs assigned by Nodes, but only if they're assigned in the object creation time. If NodeController will observe a Node object creation without the CIDR it'll start assigning one for it, and we'll have a race.

**Comment by bprashanth** [comment]
yeah i think the suggested flow is: pod-cidr doesn't make it into the node object, just into containers eth0 through CNI, nodecontroller assigns a CIDR, kubelet rewrites CNI config and somehow restarts all containers with new cidr. 

There is a less generic flow: start all hostnetwork pods but block normal pods, wait till nodecontroller assigns a CIDR, then start normal pods.

**Comment by thockin** [comment]
What if we had a "kubernetes" CNI driver for IPAM and that driver would
fail any pod until the node has a podCIDR assigned?

**Comment by bprashanth** [comment]
We can't fail *any* pod because a pod is responsible for the podCIDR, we can fail !(non-hostnetwork) pods.

The --pod-cidr option is already out there so we still need to figure out how it plays with a node.podCIDR. I guess we can say:    

1. in standalone mode --pod-cidr trumps everything
2. in non-standalone mode, only hostnetwork pods are created till node.podCIDR shows up (--pod-cidr ignored)
3. If the kubelet is started with --pod-cidr and standalone, and it creates some pods with --pod-cidr, those pods are NOT recreated even if the kubelet is restarted in !standalone mode with a node.podCIDR.

Not sure about 3

**Comment by gmarek** [comment]
1. IIUC in standalone mode Node does not register in the NC, so we don't have a problem there
2. I know absolutely nothing about Kubelets startup - is it possible to guarantee, that Kubelet won't send Node object creation before checking if it should put a podCIDR in it?

**Comment by justinsb** [comment]
Is there a workaround for 1.3 / 1.4?  What we have right now (at least in kops) is that we bring up the master with a /30 (soon to be a /29).  In kops we then allocate it a /24 PodCIDR.  But kubenet ignores this, so daemonsets (which as on 1.3 schedule onto the master) fail.  We're getting multiple user complaints.  I suspect we'll have to at least have to add this into the 1.4 issue tracker.

Some options I could try, but if anyone has any pointers that would be appreciated (e.g. "that won't work"):

1) Don't set --pod-cidr on the master.  I'm guessing this will break bootup though.
2) Set a bigger range for pod-cidr, although then these IPs won't actually work...
3) Restart kubelet once we have a podCidr (I think that will then cause kubenet to update its allocation)?

**Comment by dcbw** [comment]
Are we talking Kubenet here, or are we talking CNI?  If we're talking kubenet then --reconcile-cidr should make kubelet update the bridge address on restart at least.

But for CNI, that is all under the control of the CNI plugin itself and whatever config is on-disk when kubelet starts.  kubelet doesn't write out that configuration, and its out-of-scope for kubernetes to write that configuration out too.  The new CNI config should be deployed out-of-band when the node's podCIDR changes, or the plugin is expected to query the api-server itself for the node's new podCIDR.

And we should move kubenet to the same model when we remove the flannel/cbr0 support from kubelet.  Have kubenet watch the apiserver for the node's podCIDR and update on the fly.

**Comment by dcbw** [comment]
@justinsb if --reconcile-cidr is passed to kubelet it will watch for podCIDR changes and forward that to kubenet.  kubenet will then update the bridge's IP address and the allocation subnet.  Subsequent pods will get addresses in the new podCIDR.  Existing pods will not be re-numbered or restarted at all, but that didn't happen before either with --configure-cbr0/--reconcile-cidr AFAIK.

Why is there an initial allocation and then it's updated later?  Why not just start the node with --reconcile-cidr --network-plugin=kubenet and then allocate the full /24 whenever you're ready via the API?  Kubenet will wait to schedule pods until it gets a non-empty podCIDR.

**Comment by thockin** [comment]
When we EOL kubenet (if?) what is the story here?  It feels like CNI
actually wants to be doing the recociliation itself, and kubelet should be
totally agnostic to that, with a possible exception of CNI telling kubelet
"everything changed, restart all the pods".

**Comment by bprashanth** [comment]
@justinsb we started using kubenet in 1.3 which has this behavior (I guess you knew this as you report user frustration)

@dcbw old config-cbr0 would recreate the bridge and old docker would restart containers on daemon restart so it worked out. Right now the problem is that host ipam eats the one ip in the /30, any container created on the initial subnet gets wedged (and the /24 later allocated doesn't seem to make its way down into that container -- this may be because --pod-cidr is taking precedence in kubelet code, didn't debug). 

I don't think the kubelet will start anything, including nodecontroller which assigns cidrs, without a podcidr.

**Comment by dcbw** [comment]
@thockin yes, that would be the story here.  kubelet is totally agnostic to these kinds of network configuration games, and the plugin owns it all.  It's fine to have the apiserver store that kind of thing, but I don't think kubelet should care about it.

The one thing kubelet *does* do right now is watch that value and push it to plugins when that value changes.  Since kubenet is currently fire & forget, that wouldn't be possible for kubenet without a long-running process to watch that value, which most other more sophisticated plugins would have.  This would be a call for the UPDATE method again, perhaps where the plugin could specify on what changes it wanted UPDATE called.

**Comment by dcbw** [comment]
> @dcbw old config-cbr0 would recreate the bridge and old docker would restart containers on daemon restart so it worked out. Right now the problem is that host ipam eats the one ip in the /30, any container created on the initial subnet gets wedged (and the /24 later allocated doesn't seem to make its way down into that container -- this may be because --pod-cidr is taking precedence in kubelet code, didn't debug).

@bprashanth yes, but why is kubelet getting started with a /30 in the first place?  That seems quite pointless, since there's only 1 available address.  Why not start the node with *no* podCIDR and then set the podCIDR for that node later when you allocate the /24?

> I don't think the kubelet will start anything, including nodecontroller which assigns cidrs, without a podcidr.

kubelet would happily start, except that the network plugin will set the NodeNetworkUnavailable condition until it has a valid podCIDR. In this case as-I-understand-it, since the node *does* have a valid podCIDR (except it's a useless /30) that condition isn't set and pods get scheduled but can't actually run because there aren't an IP addresses.

The solution it seems to me is to just not set the /30 at all, but wait to assign a podCIDR until you can assign the /24, and start kubelet with --reconcile-cidr...  unless I'm totally mis-understanding the situation.  I just don't see where a /30 would be useful at all.

**Comment by dcbw** [comment]
FWIW, on sig-network a user pasted the kubelet command-line:

/usr/local/bin/kubelet --allow-privileged=true --api-servers=http://127.0.0.1:8080  --babysit-daemons=true --cgroup-root=docker --cloud-provider=aws --cluster-dns=100.64.0.10 --cluster-domain=cluster.local --config=/etc/kubernetes/manifests --configure-cbr0=true --enable-debugging-handlers=true --network-plugin=kubenet --non-masquerade-cidr=100.64.0.0/10 --pod-cidr=10.123.45.0/30 --reconcile-cidr=true --register-schedulable=false --v=2

At least two issues with that:

1. don't need --configure-cbr0 if you're using kubenet; though the code will turn off --configure-cbr0 silently to prevent issues
2. there does seem to be an initial /30 assigned to the node, which is not very helpful if anyone expects any pods to be run.  Better to just not pass --pod-cidr at all, and update the podCIDR via the apiserver when the real /24 or whatever is known.

**Comment by dcbw** [comment]
Ah, I did miss this in kubenet:

	if plugin.netConfig != nil {
		glog.V(5).Infof("Ignoring subsequent pod CIDR update to %s", podCIDR)
		return
	}

so yes, if you pass --pod-cidr on the command line, then subsequent changes will be ignored.  I believe we had a discussion about this long ago, but it's mainly to prevent the thing Tim talks about: do the old pods keep running with the old address, even though the cbr0 bridge has a new address?  That would be pointless.  There isn't any mechanism for killing the old pods or renumbering them, nor was there before AFAIK.

**Comment by dcbw** [comment]
Looks like --configure-cbr0 will change the bridge CIDR and essentially orphan old pods, and we can certainly change kubenet to do that too, though we'd need to carefully handle pod teardown to ensure the old IP addresses were released, which previously docker did for us.

**Comment by dcbw** [comment]
Reading more about this issue in https://github.com/kubernetes/kops/issues/469 and https://github.com/kubernetes/kubernetes/pull/32886 I now get why the master node has a /30 configured since it's only expected to run one pod (fluent?).  I guess if it's going to be changed to a /24 later on, why not start at /24?

**Comment by thockin** [comment]
This is sort of a disaster - can we get this fixed in a 1.4.x patch or at least early in 1.5 cycle?

**Comment by dcbw** [comment]
@thockin we certainly can, see https://github.com/kubernetes/kubernetes/issues/33132 for one take on that.  But we still need to agree on what happens to old pods that no longer have connectivity.  Do they just sit around until killed manually or what?

**Comment by freehan** [comment]
We are looking at different requirements here. 

For GKE, master is not in the same project. It is not supposed to run any user pods. No kube-proxy. No network route. Master just need a CIDR for non-hostnetwork pods. This CIDR does not need to be routable. 

For GCE, master is registered, hence it gets a podCIDR and network route, but not really used. 

For other use cases where master are supposed to run user pods, master need to get a full blown network setup plus kube-proxy.

**Comment by dcbw** [comment]
> For GKE, master is not in the same project. It is not supposed to run any user pods. No kube-proxy. No network route. Master just need a CIDR for non-hostnetwork pods. This CIDR does not need to be routable.

@freehan I don't see the distinction between user pods and other pods.  If they are scheduled by kubelet, they are pods and they get network connectivity the same as any other pod.  Can you explain the distinction more here?  Is kubelet supposed to create specific pods itself, without communication from the apiserver or something?

> For GCE, master is registered, hence it gets a podCIDR and network route, but not really used.

This should be no different than the case below, running normal pods.

> For other use cases where master are supposed to run user pods, master need to get a full blown network setup plus kube-proxy.

**Comment by dcbw** [comment]
@thockin @bprashanth @freehan It seems there are two issues mixed in this issue:

1. @justinb is talking about kops usage.  Is that relevant to this issue or should the kops thing be a separate issue?  The problem with kops is --pod-cidr is given on the kubelet command-line and thus any subsequent CIDR change is ignored.  With kubenet right now, --pod-cidr is incompatible with --reconcile-cidr; so either kops should get changed to stop passing --pod-cidr=x/30 or kubenet should be changed to allow CIDR changes.

2. fluentd failures in https://github.com/kubernetes/kubernetes/issues/32844; the cause is known and worked around by moving to a /29 and will get fixed for real by https://github.com/containernetworking/cni/pull/298.  But fluentd is normally getting run using /etc/kubernetes/manifests; is that what you mean by "non-user pods"?  Is the kubelet in the fluentd case standalone or connected to an apiserver?

**Comment by freehan** [comment]
> fluentd is normally getting run using /etc/kubernetes/manifests

Yes. That is what I mean. daemonset is a corner case, which will be fixed soon https://github.com/kubernetes/kubernetes/issues/28687

> Is the kubelet in the fluentd case standalone or connected to an apiserver?

It is in standalone mode. kubelet on GKE master do not register itself.

**Comment by freehan** [comment]
Need to settle down the requirements a bit:

For master with `register-node=false`, it goes with `pod-cidr`. 

For master with `register-node=true` and `register-schedulable=false`, it can use either `pod-cidr` or the allocated node.podCIDR. It is too bad we use pod-cidr and ignore node.podCIDR. 

If `register-node=true` and `register-schedulable=true`, then no different than a normal node.

**Comment by dcbw** [comment]
> Need to settle down the requirements a bit:
> 
> For master with register-node=false, it goes with pod-cidr.

@freehan yes, this sounds correct.

> For master with register-node=true and register-schedulable=false, it can use either pod-cidr or the allocated node.podCIDR. It is too bad we use pod-cidr and ignore node.podCIDR.

This would be against the given Flag documentation of --pod-cidr though, which talks about only being used for standalone mode.  The Flag doc can probably be changed easily enough though.

> If register-node=true and register-schedulable=true, then no difference than a normal node.

Yes.

Contrary to whatever I've said before, I think we should just remove the flags doc for --pod-cidr about standalone mode and if node.podCIDR is updated that value will be used in place of --pod-cidr *if and only if* --reconcile-cidr is also passed.

Makes it easy: you can use --pod-cidr in both standalone mode and registered mode, and then if you later update the node's podCIDR through the API the node follows that change if --reconcile-cidr is given.  What do you think?

**Comment by freehan** [comment]
> Makes it easy: you can use --pod-cidr in both standalone mode and registered mode, and then if you later update the node's podCIDR through the API the node follows that change if --reconcile-cidr is given. What do you think?

Very close to what I thought. What if we make --pod-cidr dominant? If --pod-cidr is used, node.podCIDR and reconcile-cidr will be ignored.

The only use case I can think of is bootstrapping, where master uses `pod-cidr` at first to start up and switch to node.podCIDR. We can solve it by allowing kubelet to create pods in hostnetwork when podCIDR is not configured. 

On the other hand, allowing cbr0 CIDR to change seems too complicated. That gets us into lots of reconciliation business. Is there any use case that requires this?

**Comment by justinsb** [comment]
@dcbw I described kops because I wanted to make things concrete.  It should be the same as kube-up, and thus also all installations should stay fairly close to it.

You suggested that the master shouldn't have `--pod-cidr`, but when I tested that the cluster failed to come up, because kube-controller-manager / kube-apiserver couldn't be started by kubelet.  The error is:

`Sep 23 02:15:37 ip-172-20-97-227 kubelet[771]: I0923 02:15:37.550274     771 kubelet.go:2510] skipping pod synchronization - [Kubenet does not have netConfig. This is most likely due to lack of PodCIDR]`

What am I doing wrong here @dcbw ?

The core issue that I would like to see fixed is that daemonsets have a pod scheduled onto the master, but it fails to run because it can't get a valid IP (unless it is hostNetwork).   /30 -> /29 lets us get _an_ IP, but it sounds like it won't be a _valid_ `hostNetwork: false` IP address - i.e. I don't think it will be able to talk to other pods.

As a fix, how about if kubelet performed pod synchronization before it had netConfig, but only for `hostNetwork: true` pods?  Almost by definition, all the pods that are needed to bring up enough infrastructure to hand out PodCIDRs are `hostNetwork: true`.

**Comment by caseydavenport** [comment]
>only hostnetwork pods are created till node.podCIDR shows up (--pod-cidr ignored)

Trying to catch up here and not fully groking it yet - is this suggestion just for kubenet, or CNI plugins as well?

I don't want the kubelet to enforce a dependency on node.podCIDR being defined in order to launch pods - some networks (e.g Calico) don't always care so much about that value.

**Comment by justinsb** [comment]
I tried @freehan's suggestion of allowing kubelet to run pods with hostNetwork, and it seems to "just work": https://github.com/kubernetes/kubernetes/pull/33347  Would appreciate any comments on it of course!

This approach both removes complexity and doesn't have a direct dependency on the PodCIDR - it just relies on whether the network provider considers itself ready.  So no matter why the network plugin is broken this should work.

**Comment by edevil** [comment]
I have just found this issue, after spending the last hours trying to setup a 1.4 cluster using a similar configuration that worked with 1.3.

"classic" networking was [broken](https://github.com/kubernetes/kubernetes/issues/33789) in 1.4. I used "--configure-cbr0" in 1.3 so I changed to the kubenet plugin in 1.4, otherwise docker and kubelet were both waiting for each other and never start. I'm using the Azure cloudprovider, which will fetch a pod cidr from the controller manager, but the master's kubelet starts before the controller manager so it cannot obtain a pod cidr. It just stays there with messages like:

```Kubenet does not have netConfig. This is most likely due to lack of PodCIDR```

So, bottom line, the only way I could get it to start was to specify a pod cidr at startup, which I expected to be reconciled with the controller manager when it starts. It seems this cannot be done? I also have the problem of having a fluentd pod on the master which gets an invalid IP (valid according to the bogus pod cidr I supplied at startup), the cbr0 bridge cidr is never updated.

**Comment by dcbw** [comment]
> So, bottom line, the only way I could get it to start was to specify a pod cidr at startup, which I expected to be reconciled with the controller manager when it starts. It seems this cannot be done? I also have the problem of having a fluentd pod on the master which gets an invalid IP (valid according to the bogus pod cidr I supplied at startup), the cbr0 bridge cidr is never updated.

@edevil the way kubenet is supposed to work, if you want to get whatever podCIDR the master will allocate, you don't specify one at startup, but you do pass --reconcile-cidr to kubelet.  Then when the master allocates the CIDR, kubelet will pick that up and pass to kubenet, and kubenet will create the netConfig and allow pods to start.  You need --reconcile-cidr on kubelet to make this happen though, were you passing that?

**Comment by edevil** [comment]
@dcbw I had that flag.

The whole cmd line was:
kubelet --cloud-provider=azure --cloud-config=/etc/kubernetes/azure.json --reconcile-cidr=true --api-servers=http://127.0.0.1:8080 --network-plugin-dir=/etc/kubernetes/cni/net.d --network-plugin=kubenet --register-schedulable=false --allow-privileged=true --config=/etc/kubernetes/manifests --cluster-dns=10.3.0.10 --cluster-domain=cluster.local --v 4 --node-labels=component_type=master

**Comment by dcbw** [comment]
@edevil any chance you can --v 5 and then get some kubelet logs that I can look at to figure out why it's not getting the podcidr update from the master?  What arguments is the master-controller being run with?

**Comment by edevil** [comment]
@dcbw Which "master" are we talking about here? I'm referring to the kubelet of the master node (isn't it the usual way of supervising the master components? )

Since it is the master's kubelet that needs a pod cidr to startup, it has no other components to talk to yet.

**Comment by dcbw** [comment]
@edevil ah, you mean a master kubelet where the master-controllers are run via daemonsets or something?

**Comment by edevil** [comment]
@dcbw Yes, via pod manifests in this case. I was under the impression that this was a best practice for Kubernetes clusters, instead of using systemd or something like that?

**Comment by dcbw** [comment]
@edevil can you paste in the kube-controller-manager manifest file from /etc/kubernetes/manifests?

**Comment by dcbw** [comment]
@edevil the reason I ask is because at least in upstream kube, the salt-managed kube-controller-manager.manifest has ""hostNetwork": true," in it, which should mean that kubenet isn't involved at all in running the controller-manager, and modulo a bug in kubelet or the docker/rkt runtimes, kubenet shouldn't be able to block running that manifest-defined pod due to the hostNetwork part.

**Comment by edevil** [comment]
The manifest is here:

https://github.com/edevil/kubernetes-deployment/blob/kube1.4/templates/controller-manager.yaml

In fact my whole deployment script for Kube on Azure is there.

**Comment by dcbw** [comment]
@edevil ok, so that has hostNetwork=true, which means that if the manifest pod isn't getting started, there's some bug that's not really kubenet related. And...

@thockin @freehan so kubenet can't start pods until it has a podCIDR.  It returns an error for Status() until it has one.  kubelet runtimeState records that error and then then syncLoop() checks kl.runtimeState.errors() and refuses to sync any pods because there are errors.  But a network plugin error actually shouldn't prevent a local manifested pod from starting, yet it looks like it does?

**Comment by freehan** [comment]
Yeah. There is a PR in flight for this https://github.com/kubernetes/kubernetes/pull/33347

**Comment by dcbw** [comment]
@freehan thanks, missed that PR.

**Comment by bboreham** [comment]
Looking at comments such as https://github.com/kubernetes/kubernetes/issues/32900#issuecomment-248697285 I want to add that pods are not the only reason why master might want to talk on the pod network.  It also needs that ability for urls like these:

````
  elasticsearch-logging is running at https://104.197.5.247/api/v1/proxy/namespaces/kube-system/services/elasticsearch-logging
  kibana-logging is running at https://104.197.5.247/api/v1/proxy/namespaces/kube-system/services/kibana-logging
````

where the api server will attempt to hit the service address - if `kube-proxy` isn't running it will probably fail, but if it is then the service address will translate to a pod address and so master needs to be able to route to that address.
